pkg/pmem-csi-operator/controller/deployment/controller_driver.go

```diff
@@ -198,23 +219,152 @@ func (d *PmemCSIDriver) reconcileDeploymentChanges(r *ReconcileDeployment, chang
+		klog.Warningf("Internal error while updating the annotation %s: %v", d.GetName(), err)
+	}
+
 	return false, nil
 }
 
-func (d *PmemCSIDriver) deployObjects(r *ReconcileDeployment) error {
-	objects, err := d.getDeploymentObjects()
-	if err != nil {
-		return err
+// tagOwnedObjects updates the deployment's annotation with the information(object name:GVK)
+// of the objects it owned as a json string.
+func (d *PmemCSIDriver) tagOwnedObjects(operatorVersion string) error {
+	nameGVKs := []string{}
+
+	// Nothing to update if this deployment is created by the same operator version
+	if d.ObjectMeta.Annotations[versionAnnotation] == operatorVersion {
+		return nil
 	}
-	for _, obj := range objects {
-		if err := r.Create(obj); err != nil {
+
+	for _, obj := range d.getDeploymentOwnedObjects() {
+		metaObj, err := meta.Accessor(obj)
+		if err != nil {
 			return err
 		}
+		gvk := obj.GetObjectKind().GroupVersionKind().String()
+		nameGVKs = append(nameGVKs, metaObj.GetName()+":"+gvk)
 	}
+
+	jsonString, err := json.Marshal(nameGVKs)
+	if err != nil {
+		return err
+	}
+
+	d.ObjectMeta.Annotations[versionAnnotation] = operatorVersion
+	d.ObjectMeta.Annotations[ownedObjectsAnnotation] = string(jsonString)
+	d.needsUpdate = true
+
 	return nil
 }
 
-// getDeploymentObjects returns all objects that are part of a driver deployment.
+// retrieveTaggedObjects retrives the objects tagged for this deployment by the
+// operator previously, on it's first successful reconcile
+func (d *PmemCSIDriver) retrieveTaggedObjects() ([]apiruntime.Object, error) {
+	objects := []apiruntime.Object{}
+	nameGVKs := []string{}
+
+	annotation, ok := d.ObjectMeta.Annotations[ownedObjectsAnnotation]
+	if !ok {
+		return objects, nil
+	}
+
+	if err := json.Unmarshal([]byte(annotation), &nameGVKs); err != nil {
+		return objects, err
+	}
+
+	for i := range nameGVKs {
+		parts := strings.SplitN(nameGVKs[i], ":", 2)
+		if len(parts) != 2 {
+			return nil, fmt.Errorf("failed to parse annototion %q: %s", ownedObjectsAnnotation, nameGVKs[i])
+		}
+		gvk, err := parseGroupVersionKind(parts[1])
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse annototion %q: %v", ownedObjectsAnnotation, err)
+		}
+		objects = append(objects, d.objectForGVK(gvk, parts[0]))
+	}
+
+	return objects, nil
+}
+
+// getObsoleteObjects checks if there is any change in operator version compared
+// to the one created this deployment. If there is version mismatch then it checks
+// if all the tagged objects for this deployment are still valid for the current
+// version and list all the obsolete objects that should be deleted
+func (d *PmemCSIDriver) getObsoleteObjects(currentOperatorVersion string) ([]apiruntime.Object, error) {
+	obsoleteList := []apiruntime.Object{}
+
+	deploymentVersion, ok := d.ObjectMeta.Annotations[versionAnnotation]
+	if !ok {
+		return obsoleteList, nil
+	}
+	if deploymentVersion == currentOperatorVersion {
+		return obsoleteList, nil
+	}
```

Comment:
Like the other version check this one here seems like an unnecessary optimization.